### PR TITLE
[RFC]scripts/sign.py: fix signature length calculation

### PR DESCRIPTION
**Pull request by lorc**
New versions of pycroptodome throw NotImplementedError when calling
size() method if RSA object. This is because this method is
considered insecure. As a workaround we can calculate key size
manually, in the same way as PKCS115_SigScheme does this.

Signed-off-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>

----

The following commit caused this: 

https://github.com/Legrandin/pycryptodome/commit/ab4ed2dcc1de4e96b1d9fcb63f85cdaa92396071

Also you can check https://github.com/Legrandin/pycryptodome/issues/6

This fix is a workaround, hence RFC tag. I think, proper solution would be to calculate signature and check it's real size. But in this case we have chicken and egg problem: header should contain signature size, but  it should be signed itself. We can resolve this by calculation signature for dummy data, though.

I can try to implement more proper fix is this workaround is not enough.

### CODE REVIEW

File: scripts/sign.py [hunk 1]
```diff
@@ -144,7 +144,7 @@ def main():
     h = SHA256.new()
 
     digest_len = h.digest_size
-    sig_len = ceil_div(key.size() + 1, 8)
+    sig_len = ceil_div(size(key.n) + 1, 8)
```
Comment (jenswi-linaro) [inline scripts/sign.py]: If this now is considered unsafe, perhaps we should double check this prediction matches the actual result when writing the image below in `write_image_with_signature()`

Comment (lorc) [reply scripts/sign.py]: Good idea. Will do in the next version.

**Comment by jforissier** [comment]
`PyCryptodome` != `PyCrypto` (https://pycryptodome.readthedocs.io/en/latest/src/vs_pycrypto.html). `sign.y` uses `PyCrypto`, so how could a change in `PyCryptodome` be a problem?

**Comment by lorc** [comment]
Yes, but pycrypto have two issues:
 - Last version is 2.71a, which was released 6 years ago. So it is not maintained anymore.
 - Because of this, some distributions (including Arch Linux) does not ship pycrypto

pycryptodome is replacement for pycrypto. It tries to to be compatible with pycrypto, but apparently maintainer decided to remove some old APIs.

BTW, my fix is not working as intended. I'll rework it and will come back with new version.

**Comment by jforissier** [comment]
@lorc OK, thanks. Then I think it would be sane to move to pycryptodome, but there's a catch: there are distributions-specific things to take into account. On Ubuntu for instance, a package needs to be added to the prerequisites (`apt install python[3]-cryptodome`), and the module name is `Cryptodome` not `Crypto`. I suppose this can be taken care of with a `try/except` block?

Then you may use `size_in_bits()` or `size_in_bytes()` in `RsaKey` ([doc](https://pycryptodome.readthedocs.io/en/latest/src/public_key/rsa.html#Crypto.PublicKey.RSA.RsaKey.size_in_bits)).

**Comment by lorc** [comment]
pycryptodome tries to be compatible with pycrypto APIs. So `import Crypto` works just fine.

On other hand, `size_in_bytes()` is absent in pycrypto, so I had to employ `try/except` trick.

File: scripts/sign.py [hunk 2]
```diff
@@ -124,7 +124,7 @@ def main():
     from Crypto.Signature import PKCS1_v1_5
     from Crypto.Hash import SHA256
     from Crypto.PublicKey import RSA
-    from Crypto.Util.number import ceil_div
+    from Crypto.Util.number import ceil_div, size
```
Comment (jforissier) [inline scripts/sign.py]: `size` not needed?

**Comment by lorc** [comment]
Thank you for suggestions, Jerome. I updated the commit message and removed that `size` import.

File: scripts/sign.py [hunk 3]
```diff
@@ -181,6 +187,10 @@ def sign_ta():
         else:
             signer = PKCS1_v1_5.new(key)
             sig = signer.sign(h)
+            if len(sig) != sig_len:
+                raise Exception(("Actual signature length is not equals to",
```
Comment (etienne-lms) [inline scripts/sign.py]: minor: s/equals/equal/

**Comment by lorc** [comment]
Thank you. I addressed the comments and added tags.

**Comment by jforissier** [comment]
@lorc thanks. The IBART build failure is because I have just merged the libdl stuff, so xtest expects it to be present but it's not in your branch. You can rebase if you want to leave a clean build status, or we can just ignore the failure.

**Comment by lorc** [comment]
I rebased this right now. Let's wait for the checks to be completed

**Pull request merged**